src/webnn/native/dml/DeviceDML.h

```diff
@@ -0,0 +1,63 @@
+// Copyright 2022 The WebNN-native Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef WEBNN_NATIVE_DML_DEVICEDML_H_
+#define WEBNN_NATIVE_DML_DEVICEDML_H_
+
+#include <webnn/webnn_cpp.h>
+#include <unordered_map>
+
+#include "common/Log.h"
+#include "dml_platform.h"
+#include "webnn/native/NamedOutputs.h"
+#include "webnn/native/webnn_platform.h"
+
+namespace webnn::native::dml {
+
+#define RETURN_IF_FAILED(EXPR)                                                                    \
+    do {                                                                                          \
+        auto HR = EXPR;                                                                           \
+        if (FAILED(HR)) {                                                                         \
+            dawn::ErrorLog() << "Failed to do " << #EXPR << " Return HRESULT " << std::hex << HR; \
+            return HR;                                                                            \
+        }                                                                                         \
+    } while (0)
+
+    struct DeviceDescriptor {
+        DXGI_GPU_PREFERENCE gpuPreference = DXGI_GPU_PREFERENCE_UNSPECIFIED;
+        bool useGpu = true;
+        bool useDebugLayer = false;
+    };
+
+    class Device {
```

Comment:
What do you think about replacing "Device" with something like "ExecutionContext"?

I'm pretty certain that this code originates from Dawn. In Dawn, a device object is called "logical" (meaning there is state attached to it). In D3D12, the device object  is actually stateless. The reason Dawn used a logical device is because Vulkan required such abstraction. Since DML does not depend on Vulkan, we don't need to create your own fake-logical device like this. Instead, we really just need a "context" that flushes DML commands using a queue + device + recorder.

Reply:
Yes, thanks, I agree to rename it, but there is a `ContextDM.cpp`, will we rename `Device.cpp` as `ExecutionContextDML.cpp`?  Maybe a little confused...

Reply:
Looks like we could re-use ContextDML for this. `ExecutionContextDML` could be a concrete implementation of ContextBase. We can construct a `ExecutionContextDML` by passing the D3D12 device, DML device, and D3D12 queue to it. A provider-factory, AdapterDML, would produce `ExecutionContextDML` (and creates the devices and queue). The Backend uses AdapterDML to produce this context.

Reply:
OK, I deleted `ExecutionContextDML `and moved the 
```
        ComPtr<IDMLDevice> mDevice;
        ComPtr<ID3D12Device> mD3D12Device;
        ComPtr<IDMLCommandRecorder> mCommandRecorder;
        ComPtr<ID3D12CommandQueue> mCommandQueue;
        ComPtr<ID3D12CommandAllocator> mCommandAllocator;
        ComPtr<ID3D12GraphicsCommandList> mCommandList;

        ComPtr<IDXGIAdapter1> mAdapter;
```
into ContextDML.cpp, and let Backend use `Adapter `to create  the context for DML.

Reply:
I still like ExecuteContext - "context" sounds to vague to me. But I don't feel strongly about that, you decide.